test/functional/tool_cli_completion.py

```diff
@@ -0,0 +1,290 @@
+        """ Return the max number of positional argument the option accepts. """
+        pos = list(self.arguments.keys())
+        if len(pos) == 0:
+            return 0
+
+        return max(pos)
+
+    def generate_autocomplete(self, pos):
+        """ Generate the autocomplete file line relevent to the given position pos. """
+        if len(self.arguments[pos]) == 0:
+            raise AssertionError("generating undefined arg id {} ({})".format(pos, self.arguments))
+
+        # handle special file case
+        if len(self.arguments[pos]) == 1 and len(next(iter(self.arguments[pos]))) == 0:
+            return "_filedir"
+
+        # a set order is undefined, so we order args alphabetically
+        args = list(self.arguments[pos])
+        args.sort()
+
+        return "COMPREPLY=( $( compgen -W \"" + ' '.join(args) + "\" -- \"$cur\" ) )"
+
+
+# commands where the option type can only be difficultly derived from the help message
+SPECIAL_OPTIONS = [
+        PossibleArgs("addnode").set_args(2, {"add", "remove", "onetry"}),
+        PossibleArgs("setban").set_args(2, {"add", "remove"}),
+]
+
+
+def generate_start_complete(cword):
+    """ Generate the start of an autocomplete block (beware of indentation). """
+    if cword > 1:
+        return """    if ((cword > {0})); then
+        case ${{words[cword-{0}]}} in""".format(cword)
+
+    return "    case \"$prev\" in"
+
+
+def generate_end_complete(cword):
+    """ Generate the end of an autocomplete block. """
+    if cword > 1:
+        return "\n{}esac\n{}fi\n\n".format(' '*8, ' '*4)
+
+    return "\n{}esac\n".format(' '*4)
+
+
+class CliCompletionTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_cli()
+        self.skip_if_no_wallet()
+        self.skip_if_no_bitcoind_zmq()  # zmq doesn't add commands to complete for now, but it might
+
+    def add_options(self, parser):
+        parser.add_argument(
+            '--header',
+            help='Static header part of the bitcoin-cli bash completion file',
+        )
+
+        parser.add_argument(
+            '--footer',
+            help='Static footer part of the bitcoin-cli bash completion file',
+        )
+
+        parser.add_argument(
+            '--completion',
+            help='Location of the current bash completion file',
+        )
+
+        parser.add_argument(
+            '--overwrite',
+            default=False,
+            action='store_true',
+            help='Force the test to pass by overwriting the file pointed by --completion '
+                 'to a newly generated autocompletion file',
+        )
+
+    def parse_single_helper(self, option):
+        """ Complete the arguments of option via the RPC help command. """
+
+        # if we can't find an argument list in the help text, then it has no option to add
+        regexp_res = ARG_REGEXP.match(self.nodes[0].help(option.command))
```

Comment:
```suggestion
        res = self.nodes[0].format(command=option.command, output=args_cli)
```

Instead of using regex to parse the pseudo doc-json, a hidden RPC could be added that spits out the format you want.

I think all you need to do is register a formatter in RPCMan::Check

Reply:
Yes, good idea. I have updated the branch to do that. Please have a look at the new code, thanks.

Note that some non-hidden RPC call didn't use RPCHelpMan::Check so I added a commit to fix that first (otherwise they wouldn't appear in the bash cli completion file).

Reply:
I hadn't notice #18531 when writing 412bb2f - I'll rebase this PR after #18531 is merged.